flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/AvroKeyValueSinkWriter.java

```diff
@@ -158,7 +158,7 @@ public void open(FileSystem fs, Path path) throws IOException {
 
 	@Override
 	public void close() throws IOException {
-		super.close(); //the order is important since super.close flushes inside
+		flush(); // super.close() also does a flush
 		if (keyValueWriter != null) {
```

Comment:
the writer already flushes when calling close.

If the vriter is null we should call super.close() to make sure the stream is closed.

Reply:
alternatively, we could wrap the stream we give to avro and ignore close() calls, and then close it ourselves. The we wouldn't rely on implementation details of avro.

Reply:
I would be in favour of just having
```
@Override
public void close() throws IOException {
	if (keyValueWriter != null) {
		keyValueWriter.close();
	}
}
```
then. It seems all usual stream/file stuff seems to close the stream that was handed in.

Reply:
WDYT?

Reply:
For this issue that would be enough.

However we can still leak streams if `open()` throws an exception after `super.open()` returns. Then the stream is already open. If you use the BucketingSink this will leak the stream:

```
// openNewPartFile
bucketState.writer.open(fs, inProgressPath); // if this throws exception the stream can be open
bucketState.isWriterOpen = true;
```

```
// closeCurrentPartFile
if (bucketState.isWriterOpen) { // not closing partially opened writer
	bucketState.writer.close();
	bucketState.isWriterOpen = false;
}
```

So we should modify open() to properly close the stream.

Reply:
Ah yeah, I would call super if the `keyValueWriter` is `null`. That should do it, right?

Reply:
No, because the bucketingSink never calls `close()` on the writer if `open()` threw an exception.

Reply:
Ah, so it's a pre-existing problem. I'll add a `finally` block for that.